packages/zapcode-ai-python/src/zapcode_ai/__init__.py

```
@@ -142,6 +159,39 @@ def _build_system_prompt(
+def _end_span(span: TraceSpan, status: str | None = None) -> TraceSpan:
+    span.end_time = time.time() * 1000
+    span.duration_ms = span.end_time - span.start_time
+    if status:
+        span.status = status
+    return span
+
+
+def _print_trace(span: TraceSpan, indent: int = 0) -> None:
+    prefix = "" if indent == 0 else "│ " * (indent - 1) + "├─ "
+    icon = "✗" if span.status == "error" else "✓"
+    duration = f"<1ms" if span.duration_ms < 1 else f"{span.duration_ms:.0f}ms"
```

Ruff (0.15.5)

[error] f-string without any placeholders

Remove extraneous `f` prefix

(F541)